From 5e99f612144ecc01c58658ca41d67122f63014e9 Mon Sep 17 00:00:00 2001
From: Geetha sowjanya <gakula@marvell.com>
Date: Tue, 22 Jun 2021 23:23:47 +0530
Subject: [PATCH 1594/1921] octeontx2-af: cn10k: Fix APR LMT cache flush

Current code only writes 1 to trigger entry cache flush but
need to read a 0 after writing 1 to confirm flush has been executed.
Write to 0 after a write to 1 to complete the flush. Must write to 0 before
a new flush can be issued.

This patch adds missing steps.

Change-Id: If9a9edfe663c6ea07fdc9a843fefed401c87c8bf
Signed-off-by: Geetha sowjanya <gakula@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/54685
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Devapraba Muthumani <dmuthumani@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu_cn10k.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_cn10k.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_cn10k.c
index 16ba31397e36..b89cecab1430 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_cn10k.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_cn10k.c
@@ -299,8 +299,13 @@ static int lmtst_map_table_ops(struct rvu *rvu, u32 index, u64 *val,
 		writeq((*val), (lmt_map_base + index));
 		/* Flushing the AP interceptor cache to make APR_LMT_MAP_ENTRY_S
 		 * changes effective.
+		 * First write 1 for one-time flush.
+		 * Read a 0 after writing 1 to confirm flush has been executed.
+		 * Write to 0 after a write to 1 to complete the flush.
 		 */
 		rvu_write64(rvu, BLKADDR_APR, APR_AF_LMT_CTL, BIT_ULL(0));
+		rvu_read64(rvu, BLKADDR_APR, APR_AF_LMT_CTL);
+		rvu_write64(rvu, BLKADDR_APR, APR_AF_LMT_CTL, 0x00);
 	}
 
 	iounmap(lmt_map_base);
-- 
2.31.1

